librz/bin/format/luac/luac_bin.c

```diff
@@ -0,0 +1,306 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+// SPDX-FileCopyrightText: 2021 Heersin <teablearcher@gmail.com>
+
+#include "luac_common.h"
+
+void luac_add_section(RzList *section_list, char *name, ut64 offset, ut32 size, bool is_func) {
+	RzBinSection *bin_sec = RZ_NEW0(RzBinSection);
+	if (!bin_sec) {
+		return;
+	}
+
+	bin_sec->name = rz_str_new(name);
+	bin_sec->vaddr = bin_sec->paddr = offset;
+	bin_sec->size = bin_sec->vsize = size;
+	bin_sec->add = true;
+	bin_sec->is_data = false;
+	bin_sec->bits = is_func ? sizeof(LUA_INSTRUCTION) * 8 : 8;
+	// bin_sec->has_strings = !is_func;
+	bin_sec->has_strings = false;
+	bin_sec->arch = rz_str_new("luac");
+
+	if (is_func) {
+		bin_sec->perm = RZ_PERM_R | RZ_PERM_X;
+	} else {
+		bin_sec->perm = RZ_PERM_R;
+	}
+
+	rz_list_append(section_list, bin_sec);
+}
+
+void luac_add_symbol(RzList *symbol_list, char *name, ut64 offset, ut64 size, const char *type) {
+	RzBinSymbol *bin_sym = RZ_NEW0(RzBinSymbol);
+	if (!bin_sym) {
+		return;
+	}
+
+	bin_sym->name = rz_str_new(name);
+	bin_sym->vaddr = bin_sym->paddr = offset;
+	bin_sym->size = size;
+	bin_sym->type = type;
+
+	rz_list_append(symbol_list, bin_sym);
+}
+
+void luac_add_entry(RzList *entry_list, ut64 offset, int entry_type) {
+	RzBinAddr *entry = RZ_NEW0(RzBinAddr);
+	if (!entry) {
+		return;
+	}
+
+	entry->vaddr = offset;
+	entry->paddr = offset;
+	entry->type = entry_type;
+
+	rz_list_append(entry_list, entry);
+}
+
+void luac_add_string(RzList *string_list, char *string, ut64 offset, ut64 size) {
+	RzBinString *bin_string = RZ_NEW0(RzBinString);
+	if (!bin_string) {
+		return;
+	}
+
+	bin_string->paddr = offset;
+	bin_string->vaddr = offset;
+	bin_string->size = size;
+	bin_string->length = size;
+	bin_string->string = rz_str_new(string);
+
+	rz_list_append(string_list, bin_string);
+}
+
+LuacBinInfo *luac_build_info(LuaProto *proto) {
+	rz_return_val_if_fail(proto, NULL);
+
+	LuacBinInfo *ret = RZ_NEW0(LuacBinInfo);
+	if (!ret) {
+		return NULL;
+	}
+
+	// TODO check NULL
+	ret->entry_list = rz_list_new();
```

Comment:
check nulls, as said above, and also: is it alright that no `free` method was specified? Maybe it's better to use `rz_list_newf`

Reply:
Oh these lists will be returned by `symbols` , `entries` , `sections`... would rizin free them ?

Reply:
@Heersin if you create a list with `RzList *l = rz_list_newf(some_free_function);` then when the user of this list calls `rz_list_free(l);` somewhere further in the code, the `rz_list_free()` function will call `some_free_function()` for every element. It's a useful shortcut for complex elements and allocated objects.

Reply:
Useful tips ! Thanks. I set the free function manually through `l->free = (RzListFree)free_function;` , `rz_list_newf` is a more convenient way.